prow/cmd/deck/main.go

```diff
@@ -1607,3 +1617,65 @@ func defaultLensRemoteConfig(lfc *config.LensFileConfig) error {
 
 	return nil
 }
+
+// validateStoragePath "validates" the storage bucket, *if* the `Deck.RestrictStoragePaths` config field is true.
+// The path should match a generic "<key-type>/<bucket>/<folder>/<sub-folder...>" pattern.
+// Additionally, the path's bucket should match either the built-in and default storage locations _or_
+// should be contained in the `Deck.AdditionalAllowedBuckets` config field.
+// If the `Deck.RestrictStoragePaths` field is false, no validation errors are returned (but there will be warnings logged).
+func validateStoragePath(cfg config.Getter, path string) error {
+	validate := func(cfg config.Getter, path string) error {
+		parts := strings.SplitN(path, "/", 4)
+		if len(parts) < 4 {
```

Comment:
I don't think this check is correct, for example for a 
* Presubmit, its seven elemtents: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3840/pull-ci-openshift-installer-master-images/1279069653829685248
* periodic, its four elements: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3840/pull-ci-openshift-installer-master-images/1279069653829685248
* For a batch, its six elements: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/batch/pull-ci-integr8ly-integreatly-operator-master-manifest/1279053551028408320

On top of that there are multiple strategies that can be used and that determine the structure on the object storage so IMHO its very dangerous to make any assumptions about that here

Reply:
(FYI: This part was originally Travis's change, in case you didn't know/realize.)

I _think_ you're saying that len(parts) should be somewhere between 4 and 7, depending on the type of job, but I'm not sure where you're seeing those numbers (of elements). It appears that the `SplitN(..., 4)` function limits the returned slice to _at most 4_ pieces ([golang docs](https://golang.org/pkg/strings/#SplitN)), meaning the `<sub-folder>` piece is just checked for existence.

That being said, I've personally had a bit of a problem with the pseudo-URL format of the `path` parameter -- I would have expected a "://" after the "<key-type>" (as opposed to a single "/"). I had previously noticed at least one caller (possibly the sole caller) is string-formatting the path from various pieces. I think I'd rather just pass and validate in the path as a struct, which would mean more code, but I think we'd both be satisfied.

Reply:
I looked into the [Path Strategies](https://github.com/kubernetes/test-infra/blob/088d21b2a38643ff55e59069228f449dacd4522e/prow/apis/prowjobs/v1/types.go#L529-L531) and they do have a few different ways of formatting paths. Since I'm only concerned with bucket names now, they can all be validated in the same way -- which is just "is there a folder".